File: lib/src/core/linker/resolved_metadata_cache.dart
library angular2.src.core.linker.resolved_metadata_cache;

import "../di.dart" show Injectable;
import "package:angular2/src/facade/lang.dart" show Type, isBlank;
import "element.dart" show DirectiveProvider;
import "directive_resolver.dart"
    show DirectiveResolver, CODEGEN_DIRECTIVE_RESOLVER;
import "../pipes/pipe_provider.dart" show PipeProvider;
import "pipe_resolver.dart" show PipeResolver, CODEGEN_PIPE_RESOLVER;

@Injectable()
class ResolvedMetadataCache {
  DirectiveResolver _directiveResolver;
  PipeResolver _pipeResolver;
  Map<Type, DirectiveProvider> _directiveCache =
      new Map<Type, DirectiveProvider>();
  Map<Type, PipeProvider> _pipeCache = new Map<Type, PipeProvider>();
  ResolvedMetadataCache(this._directiveResolver, this._pipeResolver) {}
  DirectiveProvider getResolvedDirectiveMetadata(Type type) {
    var result = this._directiveCache[type];
    if (isBlank(result)) {
      result = DirectiveProvider.createFromType(
          type, this._directiveResolver.resolve(type));
      this._directiveCache[type] = result;
    }
    return result;
  }

  PipeProvider getResolvedPipeMetadata(Type type) {
    var result = this._pipeCache[type];
    if (isBlank(result)) {
      result =
          PipeProvider.createFromType(type, this._pipeResolver.resolve(type));
      this._pipeCache[type] = result;
    }
    return result;
  }
}

var CODEGEN_RESOLVED_METADATA_CACHE = new ResolvedMetadataCache(
    CODEGEN_DIRECTIVE_RESOLVER, CODEGEN_PIPE_RESOLVER);

File: modules/angular2/src/core/linker/resolved_metadata_cache.ts
import {Injectable} from '../di';
import {Type, isBlank} from 'angular2/src/facade/lang';
import {DirectiveProvider} from './element';
import {DirectiveResolver, CODEGEN_DIRECTIVE_RESOLVER} from './directive_resolver';
import {PipeProvider} from '../pipes/pipe_provider';
import {PipeResolver, CODEGEN_PIPE_RESOLVER} from './pipe_resolver';

@Injectable()
export class ResolvedMetadataCache {
  private _directiveCache: Map<Type, DirectiveProvider> = new Map<Type, DirectiveProvider>();
  private _pipeCache: Map<Type, PipeProvider> = new Map<Type, PipeProvider>();

  constructor(private _directiveResolver: DirectiveResolver, private _pipeResolver: PipeResolver) {}

  getResolvedDirectiveMetadata(type: Type): DirectiveProvider {
    var result = this._directiveCache.get(type);
    if (isBlank(result)) {
      result = DirectiveProvider.createFromType(type, this._directiveResolver.resolve(type));
      this._directiveCache.set(type, result);
    }
    return result;
  }

  getResolvedPipeMetadata(type: Type): PipeProvider {
    var result = this._pipeCache.get(type);
    if (isBlank(result)) {
      result = PipeProvider.createFromType(type, this._pipeResolver.resolve(type));
      this._pipeCache.set(type, result);
    }
    return result;
  }
}

export var CODEGEN_RESOLVED_METADATA_CACHE =
    new ResolvedMetadataCache(CODEGEN_DIRECTIVE_RESOLVER, CODEGEN_PIPE_RESOLVER);
